plugins/modules/storage/zfs/zfs.py

```diff
@@ -266,13 +272,20 @@ def main():
 
     if state == 'present':
         if zfs.exists():
-            zfs.set_properties_if_changed()
+            result['diff'] = zfs.set_properties_if_changed()
         else:
             zfs.create()
+            result['diff'] = {'before': {'state': 'absent'}, 'after': {'state': state}}
```

Comment:
I'm not too familiar with Ansible's diff mode, but this seems a bit confusing. If the ZFS dataset already exists, then the keys of `diff['before']` and `diff['after']` will be the names of ZFS properties. If the ZFS dataset did not exist, then the only key will be the special string `'state'`, which is _not_ a ZFS property.

As I mentioned, I don't use diff mode too often, so maybe this is a non-issue. It just seems slightly confusing.

Reply:
It would probably be nicer to include the current properties in `diff['before']`, and an empty dict in `diff['after']` for `state=absent`, and the reverse for `state=present`. But this is also fine for me. It usually depends on how much work someone wants to invest in implementing this :)

Reply:
Sounds reasonable to me.